helix-view/src/document.rs

```diff
@@ -1034,6 +1087,39 @@ impl Default for Document {
     }
 }
 
+#[derive(Clone, Debug)]
+pub enum FormatterError {
+    SpawningFailed {
+        command: String,
+        error: std::io::ErrorKind,
+    },
+    BrokenStdin,
+    WaitForOutputFailed,
+    Stderr(String),
+    InvalidUtf8Output,
+    DiskReloadError(String),
+    NonZeroExitStatus,
+}
+
+impl std::error::Error for FormatterError {}
+
+impl Display for FormatterError {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        let s = match self {
+            Self::SpawningFailed { command, error } => {
+                format!("Failed to spawn formatter {}: {:?}", command, error)
+            }
+            Self::BrokenStdin => "Could not write to formatter stdin".to_string(),
+            Self::WaitForOutputFailed => "Waiting for formatter output failed".to_string(),
+            Self::Stderr(output) => format!("Formatter error: {}", output),
+            Self::InvalidUtf8Output => "Invalid UTF-8 formatter output".to_string(),
+            Self::DiskReloadError(error) => format!("Error reloading file from disk: {}", error),
+            Self::NonZeroExitStatus => "Formatter exited with non zero exit status:".to_string(),
+        };
+        f.write_str(&s)
```

Comment:
Rather than using `format!` and `to_string()` + `write_str`, just call `write!(f, "Could not write to formatter stdin")`, `write!(f, "Formatter error: {}", output)` etc. to avoid allocating